### DISCUSSION TITLE
Running DuckDB in Lambda functions

### DISCUSSION THREAD

**Discussion by tobilg** [duckdb/duckdb#5707] [Ideas]
Hi!

I want to run DuckDB in Lambda functions, as some companies like BoilingData and STOIC seem to already do. I'm unsure though what would be the best way to reach this goal. Node.js is my preferred language to develop serverless applications, and I mostly use the Serverless Framework with either Webpack or esbuild to package and deploy those applications to AWS.

The problems I stumbled upon so far

- The standard https://www.npmjs.com/package/duckdb package is IMO currently not usable for this, as it downloads the binaries for the host OS where the Lambda function is packaged (in my case my MacOS machine), and AWS Lambda functions run on Amazon Linux 2.
- There are very weird problems with https://github.com/mapbox/node-pre-gyp which is used to generate/publish the official DuckDB package when using a bundler like Webpack or esbuild
- I can build and package a custom DuckDB package, derived from the official package, to get rid of the `@mapbox/node-pre-gyp` issues, and create and publish a Lambda layer containing the DuckDB binary etc.
- Once I use this custom DuckDB package in my Lambda, I immediately receive a segfault error once I start querying DuckDB. Instantiation works though. I have no experience in C++ nor building custom Node.js modules, so it's pretty hard to debug from there, at least for me

If someone has done this already, I'd be very happy to learn how you did it. My current WIP can be found at https://github.com/tobilg/duckdb-lambda in case you're interested.

**Comment by tobilg** [comment]
After a lot of trial & error with esbuild & Webpack, I finally found a configuration that will correctly package the `duckdb` dependency with Webpack, but only to realize that the Linux binary that comes with the installation uses another GLIBC version than the runtime environment of Lambda functions. 

Thus seeing `/lib64/libm.so.6: version 'GLIBC_2.29' not found (required by /var/task/assets/lib/binding/duckdb.node` errors... 

So back to manually compiling DuckDB for Lambda's Amazon Linux 2 version it is.

**Reply by Mause** [reply]
You might have more luck using duckdb-wasm in unusual environments?

**Reply by tobilg** [reply]
@Mause thanks for the hint... As far as I know, duckdb-wasm has no exact feature parity with DuckDB, but that might be false. I learned that S3 globbing for example doesn't seem to work currently, and that'd be an integral feature I'd need, as well as the request waterfall outlined below.

* https://github.com/duckdb/duckdb-wasm/issues/1040
* https://github.com/duckdb/duckdb-wasm/issues/1087

**Comment by tobilg** [comment]
I managed to get it running! Now waiting for #5405 to speed up the queries for many smaller Parquet files...

**Reply by Alex-Monahan** [reply]
Nicely done! Would you be open to sharing your steps, or maybe even authoring a how to guide for our docs?

**Reply by tobilg** [reply]
@Alex-Monahan yes, I can do that, but it will take a while to write a proper documentation, and clean up my project. There's some preliminary docs at https://github.com/tobilg/duckdb-lambda though.

**Comment by iku000888** [comment]
Went through a lot of searches and found this. Want to comment that this ideally should be much easier.
Kudos to @tobilg for putting together a solution.
https://github.com/duckdb/duckdb-node/issues/15#issuecomment-2322783344

**Comment by its-felix** [comment]
For future readers, also see: https://github.com/duckdb/duckdb/discussions/17197
